python/osrd_schemas/osrd_schemas/rolling_stock.py

```diff
@@ -344,7 +362,9 @@ class RollingStock(BaseModel, extra="forbid"):
         description="The time it takes to raise this train's pantograph in s. Is null if the train is not electric.",
         default=None,
     )
-    supported_signaling_systems: List[str] = Field(default_factory=list)
+    supported_signaling_systems: list[
+        Union[SimpleSignalingSystem, EtcsSignalingSystem]
+    ] = Field(default_factory=list, discriminator="signaling_system_type")
```

Comment:
The discriminator here seems fishy. I think you would need this instead
```suggestion
    supported_signaling_systems: list[
        Annotated[Union[SimpleSignalingSystem, EtcsSignalingSystem], Field(discriminator="signaling_system_type")]
    ] = Field(default_factory=list)
```

Moving the whole `Annotated[Union[...]]` out and give it a name could be worthwhile too

Using `TypeAdapter` can also help on client's side.

Reply:
The `discriminator` is not critical because `pydantic` will just try to parse in all models in the union and pick the first that succesfully deserializes it.

But providing it ensures the right deserialization if several are actually possible, and speeds up deserializing